plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -754,58 +790,66 @@ compileExpr e = traceCompilation 2 ("Compiling expr:" GHC.<+> GHC.ppr e) $ do
-                      Just bs ->
-                        pure $ PIR.Constant annMayInline $ PLC.someValue bs
-                | GHC.getName tc == stringTyName ->
-                    case stringExprContent AsText (strip content) of
-                      Nothing ->
-                        throwSd CompilationError $
-                          "Use of fromString @BuiltinString with inscrutable content:"
-                            GHC.<+> GHC.ppr content
-                      Just bs ->
-                        case TE.decodeUtf8' bs of
-                          Right t -> pure $ PIR.Constant annMayInline $ PLC.someValue t
-                          Left err ->
-                            throwPlain . CompilationError $
-                              "Text literal with invalid UTF-8 content: " <> T.pack (show err)
-                | otherwise ->
-                    throwSd UnsupportedError $
-                      "Use of fromString on type other than builtin strings or bytestrings:"
-                        GHC.<+> GHC.ppr ty
-            Nothing ->
-              throwSd UnsupportedError $
-                "Use of fromString on type other than builtin strings or bytestrings:"
-                  GHC.<+> GHC.ppr ty
+            -- BuiltinByteString
+            Just tyCtor | GHC.getName tyCtor == builtinByteStringTyName ->
+              PIR.Constant annMayInline . PLC.someValue
+                <$> stringLiteralAsBytes builtinByteStringTyName UndoUtf8 content
+            -- BuiltinByteStringHex
+            Just tyCtor | GHC.getName tyCtor == builtinByteStringHexTyName -> do
+              hexBytes <- stringLiteralAsBytes builtinByteStringHexTyName UndoUtf8 content
+              case Base16.decode hexBytes of
+                Left err -> throwSd UnsupportedError $ "Invalid hex encoding:" GHC.<+> GHC.text err
+                Right bs -> pure $ PIR.Constant annMayInline $ PLC.someValue bs
+            -- BuiltinByteStringUtf8
+            Just tyCtor | GHC.getName tyCtor == builtinByteStringUtf8TyName ->
+              PIR.Constant annMayInline . PLC.someValue
+                <$> stringLiteralAsBytes builtinByteStringUtf8TyName PreserveUtf8 content
+            -- BuiltinString
+            Just tyCtor | GHC.getName tyCtor == stringTyName ->
+              PIR.Constant annMayInline . PLC.someValue
+                <$> stringLiteralAsText stringTyName content
+            -- For other unsupported types we have to fail compilation here,
+            -- because it won't succeed anyway:
+            -- 'fromString' function contains 'Data.Char' type in its definition
+            -- and plugin can't compile it.
+            _ -> throwUnsupported
 
     -- 'stringToBuiltinByteString' invocation
-    (strip -> GHC.Var n) `GHC.App` (strip -> stringExprContent AsBytes -> Just bs)
-      | GHC.getName n == sbbsName ->
-          pure $ PIR.Constant annMayInline $ PLC.someValue bs
+    (strip -> GHC.Var n) `GHC.App` (strip -> arg)
```

Comment:
You also need to intercept `stringToBuiltinByteStringHex`.

Reply:
Ups, good catch, totally missed it.